### DISCUSSION TITLE
can not get SSD1306 display with esp32 to work

### DISCUSSION THREAD

**Discussion by catdogmaus** [arendst/Tasmota#22104] [Q&A] [answered]
I did set my ota url to http://ota.tasmota.com/tasmota32/release/tasmota32-display.bin
I did set pins 21 and 22 in template parameters to SDA and SCL

Some tutorials suggest that it should immediately work (with some data)
Some suggest commands i2cscan , DisplayAddress 0x3c and so on
i2cscan results {"I2CScan":"Device(s) found on bus1 at 0x3c"}
However DisplayModel 2 gives {"Command":"Unknown"}
also DisplayMode 2 gives {"Command":"Unknown"}
Display remains dead

**Comment by sfromis** [comment]
You did not mention two very important points. Since Tasmota version 14, you need to configure for "Universal Display", which is indicated by configuring some unused gpio as "Option A 3", and using a display descriptor `display.ini` file to match your display. 
See https://tasmota.github.io/docs/Universal-Display-Driver/

The old display type-specific drivers in the Tasmota core are no longer used, and that's why it does not recognize the display or commands until you have configured for uDisplay, as mentioned.

**Reply by catdogmaus** [reply]
> configuring some unused gpio as "Option A 3"

What does unused mean? Unused as red gpios or ALL unused gpios. 
Did upload display.ini but no improvements so far.

**Reply by sfromis** [reply]
It means any gpio not in use, nothing else assigned to it. As the physical pin is not going to be used for "Option A 3" (which is only a flag), it *can* even be a red-marked ones, but any gpio works as well. Until you have this set up, the display will not work.

At boot, you should see messages like this, if you have enabled `Weblog 3` first.
```
00:00:00.304 I2C: Bus1 using GPIO09(SCL) and GPIO08(SDA)
...
00:00:00.371 DSP: File descriptor used
00:00:00.372 I2C: SSD1306 found at 0x3c
00:00:00.616 DSP: SSD1306 initialized
```

An easy test case is to run with `DisplayMode 1` for a simple time/date display.

**Reply by catdogmaus** [reply]
have `00:00:00.087 I2C: Bus1 using GPIO22(SCL) and GPIO21(SDA)`
that's all

**Reply by sfromis** [reply]
Well, did you make sure to have "Option A 3" configured? What's your template?

**Reply by catdogmaus** [reply]
![gpio](https://github.com/user-attachments/assets/d86e200d-641c-4b77-b5de-734a4fef0105)
![display](https://github.com/user-attachments/assets/e6825592-76bb-4348-b474-272c554f1c00)

**Reply by catdogmaus** [reply, marked as answer]
I actually figured one of the problems out, but not fully. 
All my IOT devices are denied from internet and I was not aware that I failed to give tasmota permit to access internet. It seems that when updating, but denied internet, update fails silently. (there should be some improvements possible about that behavior)
After I managed finally to update my device and screen now actually works some random toggle button appeared witch totally screws up all my rules and also setup in HA. I have no idea why it is there. I deleted all rules, Removed all relays, now I have only `Option A3`, `DS18x20`, `I2D SDA` and `I2D SCL` left. Toggle (power) button is still there. As far as I understand it switches on/off but has no effect that I could find.
